### DISCUSSION TITLE
Shipping Label endpoint (map between Receipt and shipping_label LedgerItems)

### DISCUSSION THREAD

**Discussion by morganjury** [etsy/open-api#1232] [Ideas]
Currently we are able to retrieve shipping_label_id from the ledger items endpoint, which includes the cost associated with the label. I want to be able to attribute these costs to the order that the label was created for. The receipt endpoint includes a receipt_shipping_id and tracking_code but these do not include reference to the shipping_label_id in the ledger with the cost of that label.

This is preventing me from providing accurate financial reports because shipping label costs are having to be aggregated daily rather than at the order level; shipping costs have to be attributed to the date of fulfillment rather than the order date which means profit lines spike up and down depending on orders in vs orders out on any given day (e.g. weekend orders cause profit to spike up and monday fulfillment then spikes down).

I've found this post from 2013 requesting this feature but no updates are being posted anymore and I can't see a discussion in this repo on this topic: https://groups.google.com/g/etsy-api-v2/c/R7X6ZtCzjqc?pli=1

When you inspect the order in the web app, under the shipment section (e.g. "Royal Mail 2nd Class") you can see the shipping_label_id at the bottom which aligns with the refernce id in the ledger items:
Reference                 Label ID: #231606731492

**Comment by morganjury** [comment]
There are a few places I'm smashing data together from to try and achieve this but none of it is a complete solution, currently what I am seeing is:
1. the shipments data in the receipt object contains fields for courier name and tracking ref along with a "shipped_timestamp" whose value is end of day of the date the label was purchased
2. the receipt transaction contains fields for the expected_ship_date and shipped_timestamp (shipped_timestamp here is the timestamp when the shipping label is purchased which aligns with the created_timestamp of the ledger_entry)
3. the ledger entry contains the cost of the label and the shipping_label_id

My current approach of matching the transaction.shipped_timestamp to the ledger_entry.created_timestamp works for cases where only one postage label is purchased for the order.
In cases where a second shipping label is created, the transaction timestamps are not updated. The receipt.shipments.shipped_timestamp is updated, however, I cannot use the updated receipt.shipment.shipped_timestamp since it defaults to the end of the working day which then doesn't align with the ledger_entry.created_timestamp. This means that if the wrong postage is purchased and the second label is of a different cost compared to the first label purchased then the original label value is applied to both shipments in error.

An extension to the above is that if a shipping label is cancelled/refunded then when the next shipping label is purchased THEN the transaction.shipped_timestamp is updated.

PROPOSED SOLUTION
I am hoping someone will see this request soon and implement either:
1. Add the shipping_label_id to receipt.shipments
2. Create a GET endpoint for shipping labels which responds with: shipping_label_id, ledger_entry_id, receipt_id

**Comment by jumpdumpling** [comment]
Yes! **This is vital in knowing all of the costs associated with an order.** I can get just about all the information I need regarding an order, but there is currently no way to connect a Shipping label purchased for an order under a free shipping profile. 

The only data currently available for a label purchased under free shipping profile, is in the Etsy monthly transactions.csv (SEE BELOW EXAMPLE).

Date, Type, Title, Info, Currency, Amount, Fees_Taxes, Net, Tax_Details
February 18, 2025 | Shipping | USPS shipping label | Label #262181652592 | USD | -- | -$4.14 | -$4.14 | --

I have a Label ID, but no way to connect this to an order or receipt id. None of the receipt or order details have a Label id to identify which label is associated with the order. There isn't even the amount for the label in any of the order or receipt details. I can go into a receipt in my Etsy account and pull this information from the Shipping Details Tab on the very last line "Reference Label ID: #262181652592", so there is definitely a way to identify this label with the order, however this is not available through the API.

Is there any hope of getting this information added in the order receipt or somewhere in the API? This is a very important detail and cost for managing cost of goods and choosing accurate list pricing. Thank you!

Tony

**Comment by nrathi** [comment]
+1, it would be extremely helpful for accurate reporting if the shipping label ID in the ledger had a clear correlation to the shop receipt data.

**Comment by bag06a** [comment]
Jumping on the bandwagon here....I would love to see this implemented...this NEEDS to be a feature

**Comment by ivordurham** [comment]
I have been able to find the shipping label cost for an order through the API by matching the shipped_timestamp from the receipt with the created_timestamp of a ledger entry with reference_type "shipping_label'. Specifically we look for a ledger entry's created_timestamp within 5 seconds of the receipt's shipped_timestamp. (A small sampling suggests the timestamps are identical.) We actually accumulate the ledger entries in our database, but you can take the shipped_timestamp and fetch ledger entries using that for the min_created time and, say, 5 seconds later for the max_created time. To date I've found the shipping label ledger entry uniquely. (i.e. I find exactly one shipping_label ledger entry for a given shipped_timestamp.)

This is clearly not the bullet-proof solution that a supported addition to the API would provide, but it seems to work reliably so far. Caching ledger entries in your own database makes this fairly efficient because it's not susceptible to Etsy API rate throttling. (We fetch updates once an hour, which is probably overkill but otherwise harmless.) Fetching ledger entries with a narrow min_/max_created window should also work.
